geomesa-z3/src/main/scala/org/locationtech/geomesa/curve/Z3.scala

```diff
@@ -107,52 +108,66 @@ object Z3 {
-          if (offset > 0) {
-            // some portion of this range are excluded
-            zranges(min, offset - MAX_DIM, 0, level + 1)
-            zranges(min, offset - MAX_DIM, 1, level + 1)
-            zranges(min, offset - MAX_DIM, 2, level + 1)
-            zranges(min, offset - MAX_DIM, 3, level + 1)
-            zranges(min, offset - MAX_DIM, 4, level + 1)
-            zranges(min, offset - MAX_DIM, 5, level + 1)
-            zranges(min, offset - MAX_DIM, 6, level + 1)
-            zranges(min, offset - MAX_DIM, 7, level + 1)
-            //let our children punt on each subrange
-          } else {
-            mq += (qr.min.z, qr.max.z)
-          }
+  def zranges(min: Z3, max: Z3): Seq[(Long, Long)] = {
+    val (commonPrefix, commonBits) = longestCommonPrefix(min.z, max.z)
+
+    // base our recursion on the depth of the tree that we get 'for free' from the common prefix
+    val maxRecurse = if (commonBits < 30) 7 else if (commonBits < 40) 6 else 5
+
+    val searchRange = Z3Range(min, max)
+    var mq: MergeQueue = new MergeQueue // stores our results
+
+    def zranges(prefix: Long, offset: Int, oct: Long, level: Int): Unit = {
+      val min: Long = prefix | (oct << offset) // QR + 000...
+      val max: Long = min | (1L << offset) - 1 // QR + 111...
+      val octRange = Z3Range(new Z3(min), new Z3(max))
+
+      if (searchRange containsInUserSpace octRange) {
+        // whole range matches, happy day
+        mq += (octRange.min.z, octRange.max.z)
+      } else if (searchRange overlapsInUserSpace octRange) {
+        if (level < maxRecurse && offset > 0) {
+          // some portion of this range is excluded
+          // let our children work on each subrange
+          val nextOffset = offset - MAX_DIM
+          val nextLevel = level + 1
+          zranges(min, nextOffset, 0, nextLevel)
+          zranges(min, nextOffset, 1, nextLevel)
+          zranges(min, nextOffset, 2, nextLevel)
+          zranges(min, nextOffset, 3, nextLevel)
+          zranges(min, nextOffset, 4, nextLevel)
+          zranges(min, nextOffset, 5, nextLevel)
+          zranges(min, nextOffset, 6, nextLevel)
+          zranges(min, nextOffset, 7, nextLevel)
+        } else {
+          // bottom out - add the entire range so we don't miss anything
+          mq += (octRange.min.z, octRange.max.z)
         }
-      } else if ((sr overlaps qr) || (sr overlapsInUserSpace qr)) {
-        mq += (qr.min.z, qr.max.z)
       }
     }
 
-    val prefix: Long = 0
-    val offset = MAX_BITS * MAX_DIM
-    zranges(prefix, offset, 0, 0) // the entire space
+    // kick off recursion over the narrowed space
+    zranges(commonPrefix, TOTAL_BITS - commonBits, 0, 0)
+
+    // return our aggregated results
     mq.toSeq
   }
 
   /**
+   * Calculates the longest common binary prefix between two z longs
+   *
+   * @return (common prefix, number of bits in common)
+   */
+  def longestCommonPrefix(lower: Long, upper: Long): (Long, Int) = {
```

Comment:
@jnh5y I'm fine with this routine as is, especially since it takes advantage of not checking every single bit position.  For fun, if you wanted something less intelligent (but without any conditionals), I suspect that you could use something like this:

```scala
def longestCommonPrefix(lower: Long, upper: Long) = {
  // a bit-mask in which all shared positions in the prefix are 0, and 1 everywhere else
  var v = lower ^ upper
  v = v | (v >> 1L)
  v = v | (v >> 2L)
  v = v | (v >> 4L)
  v = v | (v >> 8L)
  v = v | (v >> 16L)
  v = v | (v >> 32L)

  // count the number of bits not in the shared prefix
  var c = v - ((v >> 1L) & 0x5555555555555555L)
  c = ((c >> 2L) & 0x3333333333333333L) + (c & 0x3333333333333333L)
  c = ((c >> 4L) + c)  & 0x0F0F0F0F0F0F0F0FL
  c = ((c >> 8L) + c)  & 0x00FF00FF00FF00FFL
  c = ((c >> 16L) + c) & 0x0000FFFF0000FFFFL
  c = ((c >> 32L) + c) & 0x00000000FFFFFFFFL

  (~v, 64-c)
}
```

TL/DR:  Stick with what's already there.  It works, and is easy to read.